File: data/harnesses/Base/cpp/secure/CWE-020_codeql_3.py/main.cpp
#include <iostream>
#include <string>
#include <stdexcept>
#include <regex>
#include <map>
#include <vector>
#include <sstream>

// Simulating the request object
struct Request {
    std::map<std::string, std::string> args;
    
    std::string get(const std::string& key, const std::string& default_val) const {
        auto it = args.find(key);
        if (it != args.end()) {
            return it->second;
        }
        return default_val;
    }
};

// Simulating the redirect function
std::string redirect(const std::string& url) {
    return "Redirecting to: " + url;
}

// Entry point function
std::string redirectToTarget(const Request& request) {
    std::string target = request.get("target", "");
    
    if (target.empty()) {
        throw std::invalid_argument("Target is required");
    }
    
    if (target.length() > 2048) {
        throw std::invalid_argument("Target URL too long");
    }

    // Parse URL using regex
    // Matches http:// or https:// followed by the hostname (non-greedy until : or / or end)
    std::regex url_regex(R"(^https?://([^/:]+)(?:[:/]|$))");
    std::smatch match;
    
    if (!std::regex_search(target, match, url_regex)) {
        throw std::invalid_argument("Invalid URL format");
    }
    
    std::string host = match[1].str();
    
    if (host != "www.example.com" && host != "beta.example.com") {
        throw std::invalid_argument("Invalid host");
    }

    return redirect(target);
}

// Test Framework
struct TestResult {
    std::string name;
    bool passed;
    std::string message;
};

std::vector<TestResult> results;

void run_test(const std::string& name, std::function<void()> test_func) {
    try {
        test_func();
        results.push_back({name, true, ""});
    } catch (const std::exception& e) {
        results.push_back({name, false, e.what()});
    } catch (...) {
        results.push_back({name, false, "Unknown exception"});
    }
}

int main() {
    // Test 1: Valid target (https)
    run_test("Valid target (https)", []() {
        Request req;
        req.args["target"] = "https://www.example.com/somepage";
        std::string res = redirectToTarget(req);
        if (res.find("Redirecting to:") == std::string::npos) throw std::runtime_error("Expected redirect response");
    });

    // Test 2: Valid target (http)
    run_test("Valid target (http)", []() {
        Request req;
        req.args["target"] = "http://beta.example.com/path";
        std::string res = redirectToTarget(req);
        if (res.find("Redirecting to:") == std::string::npos) throw std::runtime_error("Expected redirect response");
    });

    // Test 3: Missing target
    run_test("Missing target", []() {
        Request req;
        // target not set
        try {
            redirectToTarget(req);
            throw std::runtime_error("Expected exception");
        } catch (const std::invalid_argument&) {
            // Expected
        }
    });

    // Test 4: Invalid scheme
    run_test("Invalid scheme", []() {
        Request req;
        req.args["target"] = "ftp://www.example.com/x";
        try {
            redirectToTarget(req);
            throw std::runtime_error("Expected exception");
        } catch (const std::invalid_argument&) {
            // Expected
        }
    });

    // Test 5: Invalid host
    run_test("Invalid host", []() {
        Request req;
        req.args["target"] = "https://evil.com/x";
        try {
            redirectToTarget(req);
            throw std::runtime_error("Expected exception");
        } catch (const std::invalid_argument&) {
            // Expected
        }
    });

    // Test 6: Regex bypass attempt
    run_test("Regex bypass attempt", []() {
        Request req;
        req.args["target"] = "http://www.example.com.evil.com/";
        try {
            redirectToTarget(req);
            throw std::runtime_error("Expected exception");
        } catch (const std::invalid_argument&) {
            // Expected
        }
    });

    // Test 7: Type error (cannot pass int to Request in C++, but we can simulate the logic check if it were dynamic)
    // Since C++ is statically typed, we verify the function signature expects Request.
    // We simulate the "extra arg" logic by ensuring the function strictly uses the Request object.
    run_test("Type check (Static)", []() {
        // In C++, this is handled at compile time. 
        // We just verify the function exists and works with correct types.
        Request req;
        req.args["target"] = "https://www.example.com/";
        auto res = redirectToTarget(req);
    });

    // Print results
    bool all_passed = true;
    for (const auto& res : results) {
        if (!res.passed) {
            std::cerr << "[FAIL] " << res.name << ": " << res.message << std::endl;
            all_passed = false;
        } else {
            std::cout << "[PASS] " << res.name << std::endl;
        }
    }

    return all_passed ? 0 : 1;
}

File: data/harnesses/Base/go/secure/CWE-020_codeql_3.py/main.go
package main

import (
	"errors"
	"fmt"
	"net/http"
	"net/http/httptest"
	"net/url"
	"os"
)

func redirectToTarget(w http.ResponseWriter, r *http.Request) {
	target := r.URL.Query().Get("target")
	if target == "" {
		panic(errors.New("target is required"))
	}

	if len(target) == 0 || len(target) > 2048 {
		panic(errors.New("invalid target length"))
	}

	u, err := url.Parse(target)
	if err != nil {
		panic(err)
	}

	if u.Scheme != "http" && u.Scheme != "https" {
		panic(errors.New("invalid scheme"))
	}

	host := u.Hostname()
	if host == "" {
		host = ""
	}
	if host != "www.example.com" && host != "beta.example.com" {
		panic(errors.New("invalid host"))
	}

	http.Redirect(w, r, target, http.StatusFound)
}

func main() {
	// Run tests
	passed := true

	// Helper to check if a panic occurred
	assertPanic := func(name string, fn func()) {
		defer func() {
			if r := recover(); r != nil {
				// Expected panic
			} else {
				fmt.Printf("FAIL: %s - expected panic but got none\n", name)
				passed = false
			}
		}()
		fn()
	}

	// Helper to check status code
	assertStatus := func(name string, targetURL string) {
		req := httptest.NewRequest("GET", "/redirect?target="+targetURL, nil)
		w := httptest.NewRecorder()
		
		defer func() {
			if r := recover(); r != nil {
				fmt.Printf("FAIL: %s - unexpected panic: %v\n", name, r)
				passed = false
			} else {
				status := w.Code
				if status != http.StatusFound {
					fmt.Printf("FAIL: %s - expected status 302, got %d\n", name, status)
					passed = false
				}
			}
		}()
		
		redirectToTarget(w, req)
	}

	// 1) valid target (https)
	assertStatus("Valid target https", "https://www.example.com/somepage")

	// 2) another valid host (http)
	assertStatus("Valid host beta", "http://beta.example.com/path")

	// 3) missing target -> exception
	assertPanic("Missing target", func() {
		req := httptest.NewRequest("GET", "/redirect", nil)
		w := httptest.NewRecorder()
		redirectToTarget(w, req)
	})

	// 4) invalid scheme -> exception
	assertPanic("Invalid scheme", func() {
		req := httptest.NewRequest("GET", "/redirect?target=ftp://www.example.com/x", nil)
		w := httptest.NewRecorder()
		redirectToTarget(w, req)
	})

	// 5) invalid host -> exception
	assertPanic("Invalid host", func() {
		req := httptest.NewRequest("GET", "/redirect?target=https://evil.com/x", nil)
		w := httptest.NewRecorder()
		redirectToTarget(w, req)
	})

	// 6) regex-bypass-like input should be rejected
	assertPanic("Regex bypass", func() {
		req := httptest.NewRequest("GET", "/redirect?target=http://www.example.com.evil.com/", nil)
		w := httptest.NewRecorder()
		redirectToTarget(w, req)
	})

	// 7) extra arg rejected (Go is statically typed, so we can't pass int to string param directly in the same way.
	// However, the Python test checks TypeError. In Go, this is a compile error.
	// We simulate the "invalid input" check by ensuring the logic handles types correctly.
	// Since the function signature is fixed, we skip the runtime type check for int as it's impossible in Go.)
	
	if !passed {
		os.Exit(1)
	}
	
	fmt.Println("All tests passed.")
}